### PR TITLE
Optimize tunnel MTU when IPv4 is used for peer

### PR DESCRIPTION
**Pull request by kubrickfr**
Per [WIREGUARD_HEADER_SIZE](https://github.com/mullvad/mullvadvpn-app/blob/087803d591f56d89681337a018622dc4760c718e/talpid-core/src/tunnel/mod.rs#L234), we assume a Header Size of 80 in all cases, but this is only a worst case scenario and should be 60 if we connect over IPv4 (see [this message from Jason A. Donenfeld](https://lists.zx2c4.com/pipermail/wireguard/2017-December/002201.html)).

---
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/3698)

### CODE REVIEW

**Review by Jontified — APPROVED** [review]
> Reviewed 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/3698)* status: :shipit: complete! all files reviewed, all discussions resolved

File: talpid-core/src/tunnel/mod.rs [hunk 1]
```diff
@@ -231,22 +231,27 @@ impl TunnelMonitor {
...
+        const WIREGUARD_HEADER_SIZE: u16 = 40;
+        let total_header_size = WIREGUARD_HEADER_SIZE + match params.connection.peer.endpoint.is_ipv6() {
+            true => IPV6_HEADER_SIZE,
+            false => IPV4_HEADER_SIZE,
+        };
+        // The largest peer MTU that we allow
+        const MAX_PEER_MTU: u16 = 1500;
         // The minimum allowed MTU size for our tunnel in IPv6 is 1280
         const MIN_IPV6_MTU: u16 = 1280;
         const MIN_IPV4_MTU: u16 = 576;
```
[the remark this answers is not in the document]
Comment (kubrickfr) [reply talpid-core/src/tunnel/mod.rs:245]: Yes, absolutely agree with all these comments, I just wanted to keep my changes minimal, but happy to make it better.

**Review by faern — APPROVED** [review]
Looks good to me. We have to test it a bit.

**Comment by kubrickfr** [comment]
This is out of scope for this PR. But I think there should be extra checks: if `MIN_IPV6_MTU > peer_mtu - total_header_size`, we should disable IPv6 inside the tunnel, and we should just fail with an appropriate error message if `MIN_IPV4_MTU > peer_mtu - total_header_size` rather than just hope for the best.

**Comment by faern** [comment]
That is indeed out of scope. This functionality was added as a best effort to correct the MTU in some cases where it was a bit off, leading to not working tunnels. I do not think we should be disabling protocols or error. At least not for now.

@Jontified Can you please test this code in all four combinations of having IPv6 on and off in the tunnel as well as connecting to a relay over v4 and v6?

**Comment by Jontified** [comment]
> That is indeed out of scope. This functionality was added as a best effort to correct the MTU in some cases where it was a bit off, leading to not working tunnels. I do not think we should be disabling protocols or error. At least not for now.
> 
> @Jontified Can you please test this code in all four combinations of having IPv6 on and off in the tunnel as well as connecting to a relay over v4 and v6?

Will do!

**Comment by kubrickfr** [comment]
Fixed in 61022499bf0407b93639bf61f4ad2d59fb7e98e8

**Comment by Jontified** [comment]
This needs to be formatted with `cargo fmt`. Could you update that?

**Comment by kubrickfr** [comment]
corde re-formatted using `cargo fmt`

**Comment by Jontified** [comment]
Merged! :tada:

**Pull request closed without merging**